README
NAME
    Encode::Arabic - Encodings of Arabic

REVISION
        $Revision: 449 $        $Date: 2007-12-10 13:46:26 +0100 (Mon, 10 Dec 2007) $

SYNOPSIS
        use Encode::Arabic;                 # imports just like 'use Encode' even with options would

        while ($line = <>) {                # renders the ArabTeX notation for Arabic both in the ..

            print encode 'utf8', decode 'arabtex', $line;          # .. Arabic script proper and the
            print encode 'utf8', decode 'arabtex-zdmg', $line;     # .. Latin phonetic transcription
        }

        # 'use Encode::Arabic ":modes"' would export the functions controlling the conversion modes

        Encode::Arabic::demode 'arabtex', 'default';
        Encode::Arabic::enmode 'buckwalter', 'full', 'xml', 'strip off kashida';

        # Arabic in lower ASCII transliterations <--> Arabic script in Perl's internal encoding

        $string = decode 'ArabTeX', $octets;
        $octets = encode 'Buckwalter', $string;

        $string = decode 'Buckwalter', $octets;
        $octets = encode 'ArabTeX', $string;

        # Arabic in lower ASCII transliterations <--> Latin phonetic transcription, Perl's utf8

        $string = decode 'Buckwalter', $octets;
        $octets = encode 'ArabTeX', $string;

        $string = decode 'ArabTeX-ZDMG', $octets;
        $octets = encode 'utf8', $string;

DESCRIPTION
    This module is a wrapper for various implementations of the encoding
    systems used for the Arabic language and covering even some non-Arabic
    extensions to the Arabic script. The included modules fit in the
    philosophy of Encode::Encoding and can be used directly with the Encode
    module.

  LIST OF ENCODINGS
    ArabTeX
        ArabTeX multi-character notation for Arabic / Perl's internal format
        for the Arabic script

        Encode::Arabic::ArabTeX, uses Encode::Mapper

    ArabTeX-RE
        Deprecated method using sequential regular-expression substitutions.
        Limited in scope over the ArabTeX notation and non-efficient in data
        processing, still, not requiring the Encode::Mapper module.

        Encode::Arabic::ArabTeX::RE

    ArabTeX-Verbatim
        ArabTeX multi-character *verbatim* notation for Arabic / Perl's
        internal format for the Arabic script

        Encode::Arabic::ArabTeX::Verbatim, uses Encode::Mapper

    ArabTeX-ZDMG
        ArabTeX multi-character notation for Arabic / Perl's internal format
        for the Latin phonetic trascription in the ZDMG style

        Encode::Arabic::ArabTeX::ZDMG, uses Encode::Mapper

    ArabTeX-ZDMG-RE
        Deprecated method using sequential regular-expression substitutions.
        Limited in scope over the ArabTeX notation and non-efficient in data
        processing, still, not requiring the Encode::Mapper module.

        Encode::Arabic::ArabTeX::ZDMG::RE

    Buckwalter
        Buckwalter one-to-one notation for Arabic / Perl's internal format
        for the Arabic script

        Encode::Arabic::Buckwalter

    Parkinson
        Parkinson one-to-one notation for Arabic / Perl's internal format
        for the Arabic script

        Encode::Arabic::Parkinson

    There are generic aliases to these provided by Encode. Case does not
    matter and all characters of the class "[ _-]" are interchangable.

    Note that the standard Encode module already deals with several other
    single-byte encoding schemes for Arabic popular with whichever operating
    system, be it *n*x, Windows, DOS or Macintosh. See Encode::Supported and
    Encode::Byte for their identification names and aliases.

  EXPORTS & MODES
    The module exports as if "use Encode" also appeared in the calling
    package. The "import" options are just delegated to Encode and imports
    performed properly, with the exception of the ":modes" option coming
    first in the list. In such a case, the following functions will be
    introduced into the namespace of the importing package:

    enmode ($enc, @list)
        Calls the "enmode" method associated with the given $enc encoding,
        and passes the @list to it. The idea is similar to the "encode"
        functions and methods of the Encode and Encode::Encoding modules,
        respectively. Used for control over the modes of conversion.

    demode ($enc, @list)
        Analogous to "enmode", but calling the appropriate "demode" method.
        See the individual implementations of the listed encodings.

SEE ALSO
    Encode::Arabic Online Interface
    <http://ufal.mff.cuni.cz/~smrz/Encode/Arabic/>

    Encode Arabic Project <http://sourceforge.net/projects/encode-arabic/>

    ElixirFM Project <http://sourceforge.net/projects/elixir-fm/>

    Klaus Lagally's ArabTeX
    <ftp://ftp.informatik.uni-stuttgart.de/pub/arabtex/arabtex.htm>

    Tim Buckwalter's Qamus <http://www.qamus.org/>

    Arabeyes Arabic Unix Project <http://www.arabeyes.org/>

    Lecture Notes on Arabic NLP
    <http://ufal.mff.cuni.cz/~smrz/ANLP/anlp-lecture-notes.pdf>

    Encode, Encode::Encoding, Encode::Mapper, Encode::Byte

    Locale::Recode

    MARC::Charset

    Lingua::AR::MacArabic, Lingua::AR::Word

    Text::TransMetaphone

AUTHOR
    Otakar Smrz, <http://ufal.mff.cuni.cz/~smrz/>

        eval { 'E<lt>' . ( join '.', qw 'otakar smrz' ) . "\x40" . ( join '.', qw 'mff cuni cz' ) . 'E<gt>' }

    Perl is also designed to make the easy jobs not that easy ;)

COPYRIGHT AND LICENSE
    Copyright 2003-2007 by Otakar Smrz

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision: 449
Author: smrz
Date: 13:46:26, December 10, 2007
Message:
Encode Arabic 1.7
----
Revision: 448
Author: smrz
Date: 23:40:47, December 8, 2007
Message:
Fixing <UW> and <uW>
----
Revision: 420
Author: smrz
Date: 10:37:55, September 26, 2007
Message:
Formal fixes
----
Revision: 419
Author: smrz
Date: 21:06:29, September 25, 2007
Message:
New releases
----
Revision: 417
Author: smrz
Date: 15:03:33, September 24, 2007
Message:
Release 1.6
----
Revision: 416
Author: smrz
Date: 11:34:49, Saturday, 22 September, 2007
Message:
Perl dependent
----
Revision: 412
Author: smrz
Date: 21:30:22, Thursday, 20 September, 2007
Message:
Build dependencies
----
Revision: 370
Author: smrz
Date: 01:01:11, Saturday, 07 July, 2007
Message:
Uninflectedness and tags
----
Revision: 368
Author: smrz
Date: 14:31:04, Thursday, 05 July, 2007
Message:
CPAN warning on non-portability
----
Revision: 343
Author: smrz
Date: 11:25:02, Wednesday, 13 June, 2007
Message:
Line-oriented interaction
----
Revision: 339
Author: smrz
Date: 02:26:53, Friday, 08 June, 2007
Message:
Fixes in roots
----
Revision: 338
Author: smrz
Date: 03:30:40, Thursday, 07 June, 2007
Message:
Morphophonemic "yY" in effect
----
Revision: 308
Author: smrz
Date: 20:58:57, Wednesday, 16 May, 2007
Message:
Improving "mikwYT"
----
Revision: 219
Author: smrz
Date: 12:34:22, Friday, 23 February, 2007
Message:
Re-organizing
----
Revision: 197
Author: smrz
Date: 15:12:25, Monday, 05 February, 2007
Message:
Extensions and improvements
----
Revision: 187
Author: smrz
Date: 14:27:14, Friday, 26 January, 2007
Message:
Portable builds
----
Revision: 179
Author: smrz
Date: 01:23:25, Sunday, 14 January, 2007
Message:
Fixing documentation
----
Revision: 166
Author: smrz
Date: 19:17:26, Thursday, 28 December, 2006
Message:
Minor changes
----
Revision: 165
Author: smrz
Date: 12:06:40, Thursday, 21 December, 2006
Message:
Defining install_base parameter
----
Revision: 162
Author: smrz
Date: 01:16:10, Saturday, 16 December, 2006
Message:
Various fixes
----
Revision: 159
Author: smrz
Date: 23:45:41, Wednesday, 13 December, 2006
Message:
svn:properties
----
Revision: 147
Author: smrz
Date: 11:21:52, Thursday, 16 November, 2006
Message:
Release 1.4
----
Revision: 146
Author: smrz
Date: 11:13:07, Thursday, 16 November, 2006
Message:
Requires with 1.4
----
Revision: 144
Author: smrz
Date: 21:18:37, Wednesday, 15 November, 2006
Message:
Finalizing documentation
----
Revision: 143
Author: smrz
Date: 02:16:57, Wednesday, 15 November, 2006
Message:
Documentation and testing
----
Revision: 142
Author: smrz
Date: 12:02:20, Tuesday, 14 November, 2006
Message:
Documentation changes
----
Revision: 128
Author: smrz
Date: 19:09:01, Saturday, 04 November, 2006
Message:
Improvements
----
Revision: 123
Author: smrz
Date: 01:59:22, Saturday, 28 October, 2006
Message:
Interfaces
----
Revision: 118
Author: smrz
Date: 13:32:09, Thursday, 19 October, 2006
Message:
Exporting and removing Encode-Korean
----
Revision: 117
Author: smrz
Date: 02:14:19, Thursday, 19 October, 2006
Message:
Separated ^^
----
Revision: 115
Author: smrz
Date: 22:27:27, Wednesday, 18 October, 2006
Message:
Fooling around
----
Revision: 114
Author: smrz
Date: 22:15:43, Wednesday, 18 October, 2006
Message:
Distributions
----

RCS file: /home/CVSROOT/smrz/perl/Encode/Arabic/Arabic.pm,v
Working file: Arabic.pm
head: 1.16
branch:
locks: strict
access list:
symbolic names:
    Encode-Arabic-1-16: 1.16
    Encode-Arabic-1-14: 1.14
    Encode-Arabic-1-13: 1.13
    Encode-Arabic-1-12: 1.12
    Encode-Arabic-1-11: 1.11
    Encode-Arabic-1-10: 1.10
    Encode-Arabic-1-09: 1.9
    start: 1.1.1.1
    smrz: 1.1.1
keyword substitution: kv
total revisions: 17;    selected revisions: 17
description:
----------------------------
revision 1.16
date: 2006/02/04 01:04:03;  author: smrz;  state: Exp;  lines: +12 -4
Encode::Arabic::Parkinson
----------------------------
revision 1.15
date: 2005/11/28 00:46:40;  author: smrz;  state: Exp;  lines: +7 -7
Incognito ^^
----------------------------
revision 1.14
date: 2005/10/02 16:08:04;  author: smrz;  state: Exp;  lines: +7 -5
Fixes in Encode::Arabic::ArabTeX
----------------------------
revision 1.13
date: 2005/07/22 20:00:02;  author: smrz;  state: Exp;  lines: +14 -5
Including Encode::Arabic::ArabTeX::Verbatim
----------------------------
revision 1.12
date: 2005/02/15 17:01:23;  author: smrz;  state: Exp;  lines: +5 -5
Extensions in Encode::Arabic::ArabTeX
----------------------------
revision 1.11
date: 2004/08/21 11:02:10;  author: smrz;  state: Exp;  lines: +80 -12
Implementation of modes
----------------------------
revision 1.10
date: 2004/01/12 22:10:28;  author: smrz;  state: Exp;  lines: +4 -4
Improvements to the installation and Encode::Arabic::ArabTeX
----------------------------
revision 1.9
date: 2003/09/08 19:43:00;  author: smrz;  state: Exp;  lines: +4 -4
Encode::Arabic Online Interface
----------------------------
revision 1.8
date: 2003/09/02 17:21:02;  author: smrz;  state: Exp;  lines: +10 -8
Buckwalter needs use ;), Encode::Arabic On-Line
----------------------------
revision 1.7
date: 2003/08/26 11:06:17;  author: smrz;  state: Exp;  lines: +28 -9
Improved documentation
----------------------------
revision 1.6
date: 2003/08/21 19:53:08;  author: smrz;  state: Exp;  lines: +103 -23
Inclusion of all modules, plus documentation
----------------------------
revision 1.5
date: 2003/08/04 09:36:02;  author: smrz;  state: Exp;  lines: +79 -79
UNIX EOL in all files
----------------------------
revision 1.4
date: 2003/07/23 13:58:59;  author: smrz;  state: Exp;  lines: +0 -0
Tests, fix of is_utf8
----------------------------
revision 1.3
date: 2003/07/21 22:48:27;  author: smrz;  state: Exp;  lines: +15 -7

lib/Encode/Arabic.pm
# ###################################################################### Otakar Smrz, 2003/01/23
#
# Encodings of Arabic ##########################################################################

# $Id: Arabic.pm 449 2007-12-10 12:46:26Z smrz $

package Encode::Arabic;

our $VERSION = '1.7' || do { q $Revision: 449 $ =~ /(\d+)/; sprintf "%4.2f", $1 / 100 };


sub import {            # perform import as if Encode were used one level before this module

    if (defined $_[1] and $_[1] eq ':modes') {

        require Exporter;

        @ISA = qw 'Exporter';
        @EXPORT_OK = qw 'enmode demode';

        __PACKAGE__->export_to_level(1, $_[0], 'enmode', 'demode');

        splice @_, 1, 1;
    }

    require Encode;

    Encode->export_to_level(1, @_);
}


use lib '..';

use Encode::Arabic::ArabTeX;
use Encode::Arabic::ArabTeX::RE;

use Encode::Arabic::ArabTeX::Verbatim;

use Encode::Arabic::ArabTeX::ZDMG;
use Encode::Arabic::ArabTeX::ZDMG::RE;

use Encode::Arabic::Buckwalter;

use Encode::Arabic::Parkinson;


sub enmode ($@) {

    my $enc = shift;
    my $obj = Encode::find_encoding($enc);

    unless (defined $obj){

        require Carp;
        Carp::croak("Unknown encoding '$enc'");
    }

    $obj->enmode(@_);
}


sub demode ($@) {

    my $enc = shift;
    my $obj = Encode::find_encoding($enc);

    unless (defined $obj){

        require Carp;
        Carp::croak("Unknown encoding '$enc'");
    }

    $obj->demode(@_);
}


1;

__END__


=head1 NAME

Encode::Arabic - Encodings of Arabic


=head1 REVISION

    $Revision: 449 $        $Date: 2007-12-10 13:46:26 +0100 (Mon, 10 Dec 2007) $


=head1 SYNOPSIS

    use Encode::Arabic;                 # imports just like 'use Encode' even with options would

    while ($line = <>) {                # renders the ArabTeX notation for Arabic both in the ..

        print encode 'utf8', decode 'arabtex', $line;          # .. Arabic script proper and the
        print encode 'utf8', decode 'arabtex-zdmg', $line;     # .. Latin phonetic transcription
    }

    # 'use Encode::Arabic ":modes"' would export the functions controlling the conversion modes

    Encode::Arabic::demode 'arabtex', 'default';
    Encode::Arabic::enmode 'buckwalter', 'full', 'xml', 'strip off kashida';

    # Arabic in lower ASCII transliterations <--> Arabic script in Perl's internal encoding

    $string = decode 'ArabTeX', $octets;
    $octets = encode 'Buckwalter', $string;

    $string = decode 'Buckwalter', $octets;
    $octets = encode 'ArabTeX', $string;

    # Arabic in lower ASCII transliterations <--> Latin phonetic transcription, Perl's utf8

    $string = decode 'Buckwalter', $octets;
    $octets = encode 'ArabTeX', $string;

    $string = decode 'ArabTeX-ZDMG', $octets;
    $octets = encode 'utf8', $string;


=head1 DESCRIPTION

This module is a wrapper for various implementations of the encoding systems used for the Arabic
language and covering even some non-Arabic extensions to the Arabic script. The included modules
fit in the philosophy of L<Encode::Encoding|Encode::Encoding> and can be used directly with the
L<Encode|Encode> module.


=head2 LIST OF ENCODINGS

=over

=item ArabTeX

ArabTeX multi-character notation for Arabic / Perl's internal format for the Arabic script

L<Encode::Arabic::ArabTeX|Encode::Arabic::ArabTeX>,
uses L<Encode::Mapper|Encode::Mapper>

=item ArabTeX-RE

Deprecated method using sequential regular-expression substitutions. Limited in scope over the
ArabTeX notation and non-efficient in data processing, still, not requiring the
L<Encode::Mapper|Encode::Mapper> module.

L<Encode::Arabic::ArabTeX::RE|Encode::Arabic::ArabTeX::RE>

=item ArabTeX-Verbatim

ArabTeX multi-character I<verbatim> notation for Arabic / Perl's internal format for the Arabic script

L<Encode::Arabic::ArabTeX::Verbatim|Encode::Arabic::ArabTeX::Verbatim>,
uses L<Encode::Mapper|Encode::Mapper>

=item ArabTeX-ZDMG

ArabTeX multi-character notation for Arabic / Perl's internal format for the Latin phonetic
trascription in the ZDMG style

L<Encode::Arabic::ArabTeX::ZDMG|Encode::Arabic::ArabTeX::ZDMG>,
uses L<Encode::Mapper|Encode::Mapper>

=item ArabTeX-ZDMG-RE

Deprecated method using sequential regular-expression substitutions. Limited in scope over the
ArabTeX notation and non-efficient in data processing, still, not requiring the
L<Encode::Mapper|Encode::Mapper> module.

L<Encode::Arabic::ArabTeX::ZDMG::RE|Encode::Arabic::ArabTeX::ZDMG::RE>

=item Buckwalter

Buckwalter one-to-one notation for Arabic / Perl's internal format for the Arabic script

L<Encode::Arabic::Buckwalter|Encode::Arabic::Buckwalter>

=item Parkinson

Parkinson one-to-one notation for Arabic / Perl's internal format for the Arabic script

L<Encode::Arabic::Parkinson|Encode::Arabic::Parkinson>

=back

There are generic aliases to these provided by L<Encode|Encode>. Case does not matter and all
characters of the class C<[ _-]> are interchangable.

Note that the standard L<Encode|Encode> module already deals with several other single-byte encoding
schemes for Arabic popular with whichever operating system, be it *n*x, Windows, DOS or Macintosh.
See L<Encode::Supported|Encode::Supported> and L<Encode::Byte|Encode::Byte> for their identification
names and aliases.


=head2 EXPORTS & MODES

The module exports as if C<use Encode> also appeared in the calling package. The C<import> options are
just delegated to L<Encode|Encode> and imports performed properly, with the exception of the
C<:modes> option coming first in the list. In such a case, the following functions will be introduced
into the namespace of the importing package:

=over

=item enmode ($enc, @list)

Calls the C<enmode> method associated with the given C<$enc> encoding, and passes the C<@list> to it.
The idea is similar to the C<encode> functions and methods of the L<Encode|Encode> and
L<Encode::Encoding|Encode::Encoding> modules, respectively. Used for control over the modes of conversion.

=item demode ($enc, @list)

Analogous to C<enmode>, but calling the appropriate C<demode> method. See the individual implementations
of the listed encodings.

=back


=head1 SEE ALSO

Encode::Arabic Online Interface L<http://ufal.mff.cuni.cz/~smrz/Encode/Arabic/>

Encode Arabic Project           L<http://sourceforge.net/projects/encode-arabic/>

ElixirFM Project                L<http://sourceforge.net/projects/elixir-fm/>

Klaus Lagally's ArabTeX         L<ftp://ftp.informatik.uni-stuttgart.de/pub/arabtex/arabtex.htm>

Tim Buckwalter's Qamus          L<http://www.qamus.org/>

Arabeyes Arabic Unix Project    L<http://www.arabeyes.org/>

Lecture Notes on Arabic NLP     L<http://ufal.mff.cuni.cz/~smrz/ANLP/anlp-lecture-notes.pdf>

L<Encode|Encode>,
L<Encode::Encoding|Encode::Encoding>,
L<Encode::Mapper|Encode::Mapper>,
L<Encode::Byte|Encode::Byte>

L<Locale::Recode|Locale::Recode>

L<MARC::Charset|MARC::Charset>

L<Lingua::AR::MacArabic|Lingua::AR::MacArabic>,
L<Lingua::AR::Word|Lingua::AR::Word>

L<Text::TransMetaphone|Text::TransMetaphone>


=head1 AUTHOR

Otakar Smrz, L<http://ufal.mff.cuni.cz/~smrz/>

    eval { 'E<lt>' . ( join '.', qw 'otakar smrz' ) . "\x40" . ( join '.', qw 'mff cuni cz' ) . 'E<gt>' }

Perl is also designed to make the easy jobs not that easy ;)


=head1 COPYRIGHT AND LICENSE

Copyright 2003-2007 by Otakar Smrz

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


=cut
